io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -997,6 +1008,20 @@ schedule thread@Thread{
           return $ Trace time tid tlbl (EventThrowTo e tid')
                  $ trace
 
+    Fix f k -> unsafeIOToST $ do
+      m <- newEmptyMVar
+      x <- unsafeInterleaveIO $
+             readMVar m
+             `catch`
+             \BlockedIndefinitelyOnMVar -> throwIO NonTermination
+      let -- recursion in 'IOSIm'
+          io = do
+            x' <- f x
+            liftST (lazyToStrictST $ unsafeIOToST (putMVar m x'))
+            return x'
+          thread' = thread { threadControl = ThreadControl (unIOSim io k) ctl }
+      unsafeSTToIO (lazyToStrictST $ schedule thread' simstate)
```

Comment:
I need more convincing that we need this low level an impl.

We're using _lazy_ ST, not strict, and we have no real concurrency. Can we not get away with a much simpler impl. Ideally using `mfix` for `ST`.